### PR TITLE
Close all caches in gunicorn postfork

### PR DESCRIPTION
**Pull request by nedbat**

### CODE REVIEW

**Comment by nedbat** [comment]
@feanil @adampalay @doctoryes opinions?

**Comment by doctoryes** [comment]
The code looks fine. :+1: 

But do we know enough about what each of these caches are? Are they all self-building? Or is there an explicit cache building step for any of them?

**Comment by nedbat** [comment]
cache.close isn't discarding any information, it's closing the connection to the cache.

**Comment by jbarciauskas** [comment]
I'd vote for explicitly closing the cache that we know is the new one, and implementing a generic solution later.

File: playbooks/roles/common_vars/defaults/main.yml [hunk 1]
```diff
@@ -140,3 +140,23 @@ common_redhat_variants:
...
+    # Close the cache so that newly forked workers cannot accidentally share
+    # the socket with the processes they were forked from. This prevents a race
+    # condition in which one worker could get a cache response intended for
+    # another worker.
+    # We do this in a way that is safe for 1.4 and 1.8 while we still have some
+    # 1.4 installations.
+    from django.conf import settings
+    from django.core import cache as django_cache
+    if hasattr(django_cache, 'caches'):
+        get_cache = django_cache.caches.__getitem__
```
Comment (adampalay) [inline playbooks/roles/common_vars/defaults/main.yml]: To be clear, this is because django will deprecate `get_cache` at some point?

**Comment by adampalay** [comment]
:+1:

File: playbooks/roles/common_vars/defaults/main.yml (hunk 1 above)
Comment (nedbat) [inline playbooks/roles/common_vars/defaults/main.yml]: Yes, and in 1.8, there is already a warning that 1.9 will get rid of get_cache.

**Comment by doctoryes** [comment]
@macdiesel  FYI

**Comment by macdiesel** [comment]
Since this is a post fork process and we are only closing the cache connections (new ones will be made in the code later) this looks good to me. :+1:

**Comment by feanil** [comment]
Tested on sandboxes. :+1:

**Comment by bradenmacdonald** [comment]
This PR seems to be breaking devstack provisioning for me - any idea why?

```
TASK: [edxapp | add gunicorn configuration files] ***************************** 
fatal: [localhost] => {'msg': "AnsibleUndefinedVariable: One or more undefined variables: 'common_close_all_caches' is undefined", 'failed': True}
fatal: [localhost] => {'msg': 'One or more items failed.', 'failed': True, 'changed': False, 'results': [{'msg': "AnsibleUndefinedVariable: One or more undefined variables: 'common_close_all_caches' is undefined", 'failed': True}]}

FATAL: all hosts have already failed -- aborting
```

I was able to work around it by explicitly making `edxapp` depend on `common_vars` but that seems like it shouldn't be necessary.

**Comment by nedbat** [comment]
@bradenmacdonald something like this happened while we were getting this right, and it turned out to be something simple, but now I can't remember what it was.  Or I'm misremembering... @feanil?

**Comment by e0d** [comment]
@bradenmacdonald is this happening on `vagrant up`?

**Comment by bradenmacdonald** [comment]
@e0d yes it was during vagrant up.

**Pull request merged**